CLOUDSTACK-10012: Towards CloudStack fat jar

rhtyd opened the pull request:
From https://github.com/apache/cloudstack/pull/2207
I've moved the branch to ASF remote to allow cross-collaboration between ACS committers. Feel free to discuss and push your changes in this branch.

- [x] Fix init.d/system.d scripts
- [x] Remove tomcat related files and config
- [x] Fix development mode
- [x] Fix CentOS6, 7, Ubuntu packaging
- [x] Validate scripts backward compatibility, run Trillian tests
- [x] Improve `ServerDaemon`
- [x] Single yml/properties file based configuration
- [x]  HTTPs/keystore management
- [x] Fix overlapping/shaded jar issues, missing classes
- [x] Allow `webapp` override, a separate UI/assets folder
- [x] Cleanup code, refactor build scripts/files
- [x] Fix docs, run upgrade tests
 
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-927

rhtyd commented:
@blueorangutan ubuntu kvm-ubuntu

rhtyd commented:
@blueorangutan test ubuntu kvm-ubuntu

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (ubuntu mgmt + kvm-ubuntu) has been kicked to run smoke tests

rhtyd commented:
Travis green, so at least the mvn side of the world works: https://travis-ci.org/apache/cloudstack/builds/261836210 :)

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

marcaurele commented:
For postmortem, there are comments on this change in the linked PR #2207

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-928

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-930

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-934

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-935

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-938

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan test centos6 kvm-centos6

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos6 mgmt + kvm-centos6) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1340)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 38902 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1340-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_network.py
Intermitten failure detected: /marvin/tests/smoke/test_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Test completed. 52 look OK, 5 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_VPC_nics_after_destroy | `Failure` | 354.75 | test_vpc_router_nics.py
test_06_router_advanced | `Failure` | 0.02 | test_routers.py
test_05_router_basic | `Failure` | 0.03 | test_routers.py
test_04_rvpc_privategw_static_routes | `Failure` | 421.53 | test_privategw_acl.py
test_09_reboot_router | `Error` | 0.10 | test_routers.py
test_08_start_router | `Error` | 40.32 | test_routers.py
test_04_restart_network_wo_cleanup | `Error` | 45.39 | test_routers.py
test_03_restart_network_cleanup | `Error` | 70.57 | test_routers.py
test_03_vpc_privategw_restart_vpc_cleanup | `Error` | 365.44 | test_privategw_acl.py
test_01_nic | `Error` | 273.45 | test_nic.py
test_reboot_router | `Error` | 379.06 | test_network.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.01 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.06 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_deploy_vgpu_enabled_vm | Skipped | 0.11 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.09 | test_deploy_vgpu_enabled_vm.py

blueorangutan commented:
<b>Trillian test result (tid-1342)</b>
Environment: kvm-ubuntu (x2), Advanced Networking with Mgmt server u
Total time taken: 39221 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1342-kvm-ubuntu.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_network.py
Intermitten failure detected: /marvin/tests/smoke/test_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Test completed. 50 look OK, 7 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_VPC_nics_after_destroy | `Failure` | 309.93 | test_vpc_router_nics.py
test_02_isolate_network_FW_PF_default_routes_egress_false | `Failure` | 219.04 | test_routers_network_ops.py
test_04_rvpc_privategw_static_routes | `Failure` | 422.02 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 240.33 | test_privategw_acl.py
test_06_verify_guest_lspci_again | `Failure` | 40.32 | test_deploy_virtio_scsi_vm.py
test_04_verify_guest_lspci | `Failure` | 40.30 | test_deploy_virtio_scsi_vm.py
ContextSuite context=TestVPCRedundancy>:setup | `Error` | 0.00 | test_vpc_redundant.py
test_01_nic | `Error` | 137.25 | test_nic.py
test_reboot_router | `Error` | 318.49 | test_network.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.01 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.04 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.05 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_deploy_vgpu_enabled_vm | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.07 | test_deploy_vgpu_enabled_vm.py

blueorangutan commented:
<b>Trillian test result (tid-1341)</b>
Environment: kvm-centos6 (x2), Advanced Networking with Mgmt server 6
Total time taken: 44621 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1341-kvm-centos6.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_list_ids_parameter.py
Intermitten failure detected: /marvin/tests/smoke/test_network.py
Intermitten failure detected: /marvin/tests/smoke/test_nic.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_router_nics.py
Test completed. 49 look OK, 8 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_VPC_nics_after_destroy | `Failure` | 403.95 | test_vpc_router_nics.py
test_06_router_advanced | `Failure` | 0.02 | test_routers.py
test_05_router_basic | `Failure` | 0.02 | test_routers.py
test_04_rvpc_privategw_static_routes | `Failure` | 482.04 | test_privategw_acl.py
test_05_change_vm_ostype_restart | `Failure` | 22.39 | test_deploy_virtio_scsi_vm.py
test_03_verify_libvirt_attach_disk | `Failure` | 12.09 | test_deploy_virtio_scsi_vm.py
test_02_verify_libvirt_after_restart | `Failure` | 143.78 | test_deploy_virtio_scsi_vm.py
test_01_verify_libvirt | `Failure` | 1.06 | test_deploy_virtio_scsi_vm.py
test_09_reboot_router | `Error` | 5.12 | test_routers.py
test_08_start_router | `Error` | 45.35 | test_routers.py
test_04_restart_network_wo_cleanup | `Error` | 45.41 | test_routers.py
test_03_restart_network_cleanup | `Error` | 85.62 | test_routers.py
test_03_vpc_privategw_restart_vpc_cleanup | `Error` | 291.17 | test_privategw_acl.py
test_01_nic | `Error` | 187.81 | test_nic.py
test_reboot_router | `Error` | 343.63 | test_network.py
ContextSuite context=TestListIdsParams>:setup | `Error` | 0.00 | test_list_ids_parameter.py
test_01_internallb_roundrobin_1VPC_3VM_HTTP_port80 | `Error` | 122.51 | test_internal_lb.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.01 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.03 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_02_list_snapshots_with_removed_data_store | Skipped | 0.00 | test_snapshots.py
test_01_snapshot_root_disk | Skipped | 0.00 | test_snapshots.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.03 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_deploy_vgpu_enabled_vm | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py

marcaurele commented:
I came across another change that is required for the fatjar to work. We have to change the DbUpgrade process to not load `File` but return an array of `InputStream` since we have to load those files from inside the JAR.

marcaurele commented:
@rhtyd The DB files are kept out of the jar. Any opinion on that one if we should keep them outside (due to some possible customization for providers/plugins) or should we include them in the `META-INF` directory of the jar?

rhtyd commented:
@marcaurele I think we need to consider and discuss three things;
- Whether to bundle the UI within the jar, or provider a webapp path to say /usr/share/cloudstack-management/webapp, in this directory we copy the UI asset files. This would allow admins to put their customizations. A solution could be to do both, so the fat jar has the UI files and however `ServerDaemon` would set the webapp path to the /usr path by default (if available) and fallback to assets in the jar when not found.
- The DB files are kept out of the jar in case someone want to fix any upgrade failure issue. It's kind of subjective. My opinion is to bundle them in the jar, which reduces file management and opens scope for changes to the sql paths.
- In addition, a lot of people may want to consume additional jars (plugins etc) which is why bounty castle and other jars may be put at `/usr/share/cloudstack-management/lib`. When mgmt server starts, we scan and include all the jars at this path in the classpath, with them first in the order in case someone want to overide any bundled jar etc.

marcaurele commented:
@rhtyd 

1. I will add a startup parameter `-Dwebappdir=...` to let the user change the path location instead of trying to guess things. Then if the path is incorrect at least it will return an error instead of loading the default UI. IMO it's better.
2. I'm in favor to bundle them inside the JAR. It's best for people customizing the installation to have an extra module where they do what they need, instead of editing files from the official releases.
3. +1 for the JARs ordering, that make sense for customization of the install with extra jars.

I'll push some of the changes today.

marcaurele commented on packaging/centos63/cloud-management.rc:
```diff
@@ -1,116 +1,153 @@
+setJavaHome
+
+JARS=$(ls /usr/share/cloudstack-management/lib/*.jar | tr '\n' ':' | sed s'/.$//')
+CLASSPATH="/etc/cloudstack/management:/usr/share/cloudstack-common:/usr/share/cloudstack-management/setup:/usr/share/cloudstack-management:/usr/share/cloudstack-management/cloud-management-server.jar:$JARS"
```
`:$JARS` should be put in front here to first load extra JARs

marcaurele commented:
@rhtyd I cannot push to this branch

rhtyd commented:
@marcaurele perhaps you don't have commit rights yet, can you send a PR to the asf branch `CLOUDSTACK-10012` and ping me I'll merge your changes right away

marcaurele commented:
Related to 42d9297178e6d8cd3fbfc1785b8905cc7e99534b, I forgot to remove the `jetty.` in the class javadoc of `ServerDaemon`- if you could do it, thanks (PR seems overkill)

rhtyd commented:
Thanks @marcaurele, sure will do that. These days I'm down with some £dayjob project work but I've planned to resume working on this PR after next ~2 weeks.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

rhtyd commented:
@marcaurele I've resumed working on this now. I've squashed changes by authors, after painfully going through 27+ commits while rebasing against the latest master, hope you don't mind. I did not see any classes being overwritten for bountycastle and hence removed it from the exclusion list, do you think that could cause any issues? I'm exploring some ways to further improve the `ServerDaemon` launcher by using a customer properties or yml file that can be used as a single point to configure ports, listening interfaces, https, keystore etc.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-1162

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1163

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1164

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-1166

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

marcaurele commented:
@rhtyd You must exclude bountycastle due to the Java security mechanism that validates all JCE providers jar, to avoid manipulating those. Otherwise you'll get an exception when reaching the code that uses bouncycastle lib `JCE cannot authenticate the provider BC`

rhtyd commented:
@marcaurele I tried running marvin tests (CA tests around security, cert generation etc) they are passing and I saw no failures or exceptions, can you suggest a way to reproduce the error?

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-1167

blueorangutan commented:
<b>Trillian test result (tid-1593)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 29266 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1593-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 59 look OK, 4 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 45.77 | test_vpc_vpn.py
test_05_rvpc_multi_tiers | `Failure` | 181.46 | test_vpc_redundant.py
test_04_rvpc_privategw_static_routes | `Failure` | 294.95 | test_privategw_acl.py
test_05_add_annotation_for_invvalid_entityType | `Error` | 0.04 | test_host_annotations.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.02 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.07 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.01 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.01 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.01 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.01 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.01 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.01 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.01 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.01 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.01 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.01 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.01 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.01 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.01 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.01 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.01 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.02 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py

marcaurele commented:
@rhtyd I tried on my side too and I'm not able to reproduce the exception. I must have corrected the exclusions list of the shaded jar afterwards.

rhtyd commented:
@marcaurele you may see list of shaded jars included/excluded from the build output, the excluded list currently has few jars. I checked the bountycastle jar is shaded without including the keys from META-INF (signed content), perhaps this is why we're not seeing the exception. With this, do you still think this may cause issues in future and we should not shade the bountycastle jars?

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✔debian. JID-1168

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

marcaurele commented:
IMO it's cleaner to keep bouncycastle out of the shaded jar, to keep the JCE security checks.

rhtyd commented:
@marcaurele I agree, I've already made changes to put the bcprov jars in a separate `lib` folder that's included in the class path. The only external `jar` dependency is the distro-provided `mysql-connector-java`, due to licencing issues for normal builds this dependency cannot be shaded with the uberjar. 
Other changes I did so far are:
- Remove unnecessary dependencies, reduce uber jar size
- Work-in-progress single yml file based configuration, instead of passing flags to `ServerDaemon`
- In-built webapp UI assets in uberjar, and additional editable webapp folder at `/usr/share/cloudstack-management` for UI plugin installation, UI edits etc
- WIP SSL keystore/cert management

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1169

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1175

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1176

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-1177

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1178

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1180

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1181

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1183

rhtyd commented:
This is ready for review now, I've personally tested this on CentOS6, CentOS7, Ubuntu 14.04 and 17.04. Pinging for review -- @marcaurele @wido @PaulAngus @DagSonstebo @DaanHoogland @nvazquez @borisstoyanov @resmo @karuturi @fmaximus @GabrielBrascher @rafaelweingartner @ustcweizhou @NuxRo 

New configuration file updates/changes:
- `/etc/cloudstack/management/server.properties` - single file to configure ports, webapp directory, http/s configurations, webapp context path etc.
- Environment configuration set at `/etc/default/cloudstack-management` (CentOS accepts having env conf files at /etc/default, while Ubuntu uses this path most of the time)

High-level changelog:

    - Changes packaging to shaded jar
    - Updates jetty dependencies version 9.2 (java 7+)
    - Initial changes to debian config files
    - Initial removal of tomcat dependency
    - Introduce ServerDaemon class with customizations
    - Migrate to embedded Jetty server
    - Improve ServerDaemon implementation
    - Introduce a new server.properties file for easier configuration
    - Have a single /etc/default/cloudstack-management to configure env
    - Reduce shaded jar file, removing unnecessary dependencies
    - Upgrade to Spring 5.x, upgrade several jar dependencies
    - Does not shade and include mysql-connector, used from classpath instead
    - Upgrade and use bountcastle as a separate un-shaded jar dependency
    - Remove tomcat related configuration and files
    - Have both embedded UI assets in uber jar and separate webapp directory
    - Refactor systemd and init scripts, cleanup packaging
    - Made cloudstack-setup-databases faster, using `urandom`
    - Remove unmaintained distro packagings

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1184

rhtyd commented:
@blueorangutan test ubuntu kvm-ubuntu

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (ubuntu mgmt + kvm-ubuntu) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan test centos6 vmware-55u3

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos6 mgmt + vmware-55u3) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1601)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 35610 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1601-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 60 look OK, 3 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 81.19 | test_vpc_vpn.py
test_04_rvpc_privategw_static_routes | `Failure` | 329.39 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 138.14 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 234.00 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 61.81 | test_privategw_acl.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.07 | test_host_annotations.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.04 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.10 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 1.07 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.04 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.05 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.04 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.05 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.05 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.04 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.05 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.05 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.04 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.04 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.04 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.04 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py

blueorangutan commented:
<b>Trillian test result (tid-1600)</b>
Environment: kvm-ubuntu (x2), Advanced Networking with Mgmt server u
Total time taken: 36342 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1600-kvm-ubuntu.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 57 look OK, 6 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 66.14 | test_vpc_vpn.py
test_07_resize_fail | `Failure` | 15.48 | test_volumes.py
test_04_rvpc_privategw_static_routes | `Failure` | 315.15 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 153.85 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 305.86 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 87.66 | test_privategw_acl.py
test_06_verify_guest_lspci_again | `Failure` | 40.55 | test_deploy_virtio_scsi_vm.py
test_04_verify_guest_lspci | `Failure` | 40.95 | test_deploy_virtio_scsi_vm.py
test_02_create_template_with_checksum_sha1 | `Error` | 66.06 | test_templates.py
test_02_create_template_with_checksum_sha1 | `Error` | 71.20 | test_templates.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.08 | test_host_annotations.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.03 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.08 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.12 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 1.13 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.04 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.03 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.04 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.03 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_kvm_host_degraded | Skipped | 31.49 | test_hostha_kvm.py
test_deploy_vgpu_enabled_vm | Skipped | 0.03 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py

blueorangutan commented:
<b>Trillian test result (tid-1602)</b>
Environment: vmware-55u3 (x2), Advanced Networking with Mgmt server 6
Total time taken: 46994 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1602-vmware-55u3.zip
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 58 look OK, 5 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_vpc_remote_access_vpn | `Failure` | 131.47 | test_vpc_vpn.py
test_01_create_volume | `Failure` | 159.53 | test_volumes.py
test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false | `Failure` | 455.88 | test_routers_network_ops.py
test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true | `Failure` | 440.90 | test_routers_network_ops.py
test_04_rvpc_privategw_static_routes | `Failure` | 682.43 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 369.78 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 380.35 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 142.60 | test_privategw_acl.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.06 | test_host_annotations.py
test_08_resize_volume | Skipped | 5.24 | test_volumes.py
test_07_resize_fail | Skipped | 10.34 | test_volumes.py
test_09_copy_delete_template | Skipped | 0.01 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 66.50 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.08 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.03 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.02 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.02 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.02 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_remove_ha_provider_not_possible | Skipped | 0.03 | test_hostha_kvm.py
test_hostha_kvm_host_recovering | Skipped | 0.05 | test_hostha_kvm.py
test_hostha_kvm_host_fencing | Skipped | 0.03 | test_hostha_kvm.py
test_hostha_kvm_host_degraded | Skipped | 0.04 | test_hostha_kvm.py
test_hostha_enable_ha_when_host_in_maintenance | Skipped | 0.03 | test_hostha_kvm.py
test_hostha_enable_ha_when_host_disconected | Skipped | 0.03 | test_hostha_kvm.py
test_hostha_enable_ha_when_host_disabled | Skipped | 0.03 | test_hostha_kvm.py
test_hostha_configure_default_driver | Skipped | 0.04 | test_hostha_kvm.py
test_disable_oobm_ha_state_ineligible | Skipped | 0.03 | test_hostha_kvm.py
test_06_verify_guest_lspci_again | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_05_change_vm_ostype_restart | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_04_verify_guest_lspci | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_03_verify_libvirt_attach_disk | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_02_verify_libvirt_after_restart | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_01_verify_libvirt | Skipped | 0.00 | test_deploy_virtio_scsi_vm.py
test_deploy_vgpu_enabled_vm | Skipped | 1.35 | test_deploy_vgpu_enabled_vm.py

DaanHoogland commented:
@rhtyd @marcaurele thanks to you both. We might just need a tad more testing but LGTM otherwise

marcaurele commented:
LGTM - thanks @rhtyd for all the extra work

rhtyd commented:
@borisstoyanov thanks, with embedded Jetty until the management server starts the embedded Jetty server does not start to listen on http/https ports which is why you get connection refused quite promptly.
Have you also performed in-place upgrade tests, i.e. say upgrading from 4.9.3 to 4.11/fatjar? We can leave reviewing other PRs until this is well tested and merged first. Thanks again.

rhtyd commented:
@borisstoyanov I think there is still some additional work left in #2247, this PR is not dependent on #2247, so we should test in-place upgrades from 4.9/4.10 to rpms/debs from this PR. Thanks, looking forward to your test results and comments.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1188

wido commented:
Why are we using JSVC in the systemd scripts? I know we need this in the init.d/sysvinit scripts, but it's not required by systemd.

Any reason for that?

rhtyd commented:
@wido we want to use the same rpm/deb build scripts and packages to work which is why I've made them to use jsvc for now. For example, this allows the same deb pkg to work on Ubuntu14.04+older versions.

I refactored the systemd service scripts for both centos and ubuntu -- this allows the systemd scripts to be same for all systemd enabled distros, however in my testing against centos, I found that `ExecStartPre=/bin/bash -c "/bin/systemctl set-environment JARS=$(ls /usr/share/cloudstack-management/lib/*.jar | tr '\n' ':' | sed s'/.$//')"` did not work as expected when user was `cloud` and `Type=simple`, so I changed it to run as root user which can set the env variable and use jsvc to fork the child process to run as `cloud` user.  I've also noted in the limitation/future work that we may want to get rid of this in future (i.e. move to systemd+java directly).

wido commented:
Understood @rhtyd ! Thanks for the feedback

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1189

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

wido commented:
I just checked, building DEB files worked flawlessly. cloudstack.jar is 73M in size. Good!

rhtyd commented:
Thanks @wido, is that a LGTM on this PR?

blueorangutan commented:
<b>Trillian test result (tid-1606)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 41902 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2226-t1606-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_host_annotations.py
Intermitten failure detected: /marvin/tests/smoke/test_iso.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Test completed. 54 look OK, 9 have error(s)


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_07_resize_fail | `Failure` | 15.34 | test_volumes.py
test_10_attachAndDetach_iso | `Failure` | 1513.31 | test_vm_life_cycle.py
test_09_destroy_ssvm | `Failure` | 0.04 | test_ssvm.py
test_07_reboot_ssvm | `Failure` | 0.04 | test_ssvm.py
test_05_stop_ssvm | `Failure` | 0.04 | test_ssvm.py
test_03_ssvm_internals | `Failure` | 0.03 | test_ssvm.py
test_01_list_sec_storage_vm | `Failure` | 0.04 | test_ssvm.py
test_04_rvpc_privategw_static_routes | `Failure` | 223.27 | test_privategw_acl.py
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 117.63 | test_privategw_acl.py
test_02_vpc_privategw_static_routes | `Failure` | 117.59 | test_privategw_acl.py
test_01_vpc_privategw_acl | `Failure` | 46.57 | test_privategw_acl.py
test_04_extract_Iso | `Failure` | 5.08 | test_iso.py
ContextSuite context=TestVpcSite2SiteVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVpcRemoteAccessVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestRVPCSite2SiteVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVPCRedundancy>:setup | `Error` | 0.00 | test_vpc_redundant.py
test_06_download_detached_volume | `Error` | 20.28 | test_volumes.py
test_05_create_template_with_no_checksum | `Error` | 65.61 | test_templates.py
test_04_extract_template | `Error` | 5.09 | test_templates.py
test_04_create_template_with_checksum_md5 | `Error` | 65.51 | test_templates.py
test_03_delete_template | `Error` | 5.09 | test_templates.py
test_03_create_template_with_checksum_sha256 | `Error` | 65.51 | test_templates.py
test_02_create_template_with_checksum_sha1 | `Error` | 65.48 | test_templates.py
test_01_create_template | `Error` | 65.66 | test_templates.py
test_05_add_annotation_for_invalid_entityType | `Error` | 0.07 | test_host_annotations.py
test_change_service_offering_for_vm_with_snapshots | Skipped | 0.00 | test_vm_snapshots.py
test_09_copy_delete_template | Skipped | 0.02 | test_templates.py
test_06_copy_template | Skipped | 0.00 | test_templates.py
test_static_role_account_acls | Skipped | 0.02 | test_staticroles.py
test_11_ss_nfs_version_on_ssvm | Skipped | 0.02 | test_ssvm.py
test_01_scale_vm | Skipped | 0.00 | test_scale_vm.py
test_01_primary_storage_iscsi | Skipped | 0.09 | test_primary_storage.py
test_vm_nic_adapter_vmxnet3 | Skipped | 0.00 | test_nic_adapter_type.py
test_03_nic_multiple_vmware | Skipped | 1.06 | test_nic.py
test_nested_virtualization_vmware | Skipped | 0.00 | test_nested_virtualization.py
test_06_copy_iso | Skipped | 0.00 | test_iso.py
test_list_ha_for_host_valid | Skipped | 0.04 | test_hostha_simulator.py
test_list_ha_for_host_invalid | Skipped | 0.04 | test_hostha_simulator.py
test_list_ha_for_host | Skipped | 0.03 | test_hostha_simulator.py
test_hostha_enable_feature_without_setting_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_enable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_disable_feature_valid | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_invalid_provider | Skipped | 0.02 | test_hostha_simulator.py
test_hostha_configure_default_driver | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_recovering | Skipped | 0.03 | test_hostha_simulator.py
test_ha_verify_fsm_fenced | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_degraded | Skipped | 0.02 | test_hostha_simulator.py
test_ha_verify_fsm_available | Skipped | 0.03 | test_hostha_simulator.py
test_ha_multiple_mgmt_server_ownership | Skipped | 0.03 | test_hostha_simulator.py
test_ha_list_providers | Skipped | 0.03 | test_hostha_simulator.py
test_ha_enable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_disable_feature_invalid | Skipped | 0.02 | test_hostha_simulator.py
test_ha_configure_enabledisable_across_clusterzones | Skipped | 0.02 | test_hostha_simulator.py
test_configure_ha_provider_valid | Skipped | 0.03 | test_hostha_simulator.py
test_configure_ha_provider_invalid | Skipped | 0.03 | test_hostha_simulator.py
test_deploy_vgpu_enabled_vm | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py
test_3d_gpu_support | Skipped | 0.04 | test_deploy_vgpu_enabled_vm.py

borisstoyanov commented:
@rhtyd after upgrading from 4.5 I wasn't able to start management server, got the following: 
```
[root@trl-1020-v-cs45-bstoyanov-mgmt1 ~]# service cloudstack-management restart
ls: cannot access /usr/share/cloudstack-management/lib/*.jar: No such file or directory
Stopping CloudStack Management Server cloudstack-management[  OK  ]
cat: /var/run/cloudstack-management.pid: No such file or directory]
                                                           [FAILED]
```

rhtyd commented:
Rebased to latest master. This is ready for merge now.
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

rhtyd commented:
@borisstoyanov okay I'll look into it

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1201

wido commented:
Yes, LGTM from me.

I tested a clean install on Ubuntu 16.04 with MySQL 5.7 ( https://blog.widodh.nl/2017/04/apache-cloudstack-and-mysql-5-7/ )

May I however ask that in *server.properties*  this is set:

<pre>
# The binding interface used by the management server
bind.interface=::
</pre>

This will make sure that the server also listens on IPv6

wido commented on client/conf/server.properties.in:
```diff
@@ -0,0 +1,41 @@
+# under the License.
+
+# The binding interface used by the management server
+bind.interface=0.0.0.0
```
Can we change *0.0.0.0* to *::* so that it listens on IPv6 by default?

I tested this on Ubuntu 16.04 and works fine:

<pre>
root@cloudstack-man-test:~# curl -I http://127.0.0.1:8080/client/
HTTP/1.1 200 OK
Content-Type: text/html
ETag: W/"114557-1499056602000
Content-Length: 114557
Server: Jetty(9.2.z-SNAPSHOT)

root@cloudstack-man-test:~# curl -I http://[::1]:8080/client/
HTTP/1.1 200 OK
Content-Type: text/html
ETag: W/"114557-1499056602000
Content-Length: 114557
Server: Jetty(9.2.z-SNAPSHOT)

root@cloudstack-man-test:~#
</pre>

rhtyd replied:
Hi @wido I think the standard I've seen in mysql and elsewhere is to use this. It's a matter of config issue, and ipv6 may not be enabled on all systems.

For me it fails when JAVA_OPT has `-Djava.net.preferIPv4Stack=true` set, and bind.interface has `::` set. Exception seen is:
```
WARN  [o.e.j.u.c.AbstractLifeCycle] (main:null) (logid:) FAILED ServerConnector@4e9ba398{HTTP/1.1}{:::8080}: java.net.SocketException: Protocol family unavailable
java.net.SocketException: Protocol family unavailable
	at sun.nio.ch.Net.bind0(Native Method)
```
Command ran to test this case (build cloudstack and cd to client/target, fix conf/server.properties): `java -Dlog4j.configuration=conf/log4j-cloud.xml -cp cloud-client-ui-4.11.0.0-SNAPSHOT.jar:lib/bcpkix-jdk15on-1.58.jar:lib/bcprov-jdk15on-1.58.jar:.:/usr/share/java/mysql-connector-java.jar:conf:common:utilities/scripts/db/ -Djava.net.preferIPv4Stack=true  org.apache.cloudstack.ServerDaemon`

wido replied:
Yes, that's true. However, all modern Linux systems have IPv6 enabled. It's a manual action to disable IPv6 on a system.

You don't need a public IPv6 address though to listen on *::*, you already have your localhost *::1* to listen on.

Don't forget that the *current* Tomcat6/7 already listen on IPv6 by default. So it would be a change from 4.10 to 4.11 if we wouldn't listen on IPv6 anymore.

rhtyd replied:
@wido I agree with you, however, for the default I needed to put some configuration that would work out of the box on most/all systems. I know several installations where they have put `-Djava.net.preferIPv4Stack=true ` in their global JAVA opts (or in cloudstack specific env/init/systemd files). That said, we can have a compromise where I can remove the default value set to empty, i.e. `bind.interface=`, that way the system picks up the default interface that could work automatically.

With this approach, on systems without ipv6, this listens on:
tcp        0      0 127.0.0.1:8080          0.0.0.0:*               LISTEN     

And with ipv6 enabled, it listens on:
tcp6       0      0 127.0.0.1:8080          :::*                    LISTEN

rhtyd replied:
@wido check now, I've made the above change.

rhtyd commented:
Thanks for testing and reviewing @wido.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1203

rhtyd commented:
I've confirmed manual upgrade (in-place) testing, from ACS 4.9.3.0 -> 4.11.0.0-SNAPSHOT. Worked as expected, other than seeding systemvmtemplates, the upgrade worked without issues, the old tomcat files are not removed. I'll update upgrade/release notes on the same after this PR gets merged.

rhtyd commented:
LGTM, merging based on reviews, tests and upgrade-tests.
